File: ios/FreakCore/Sources/FreakCore/Device.swift
// Device.swift — FreakDeviceProtocol (the app-facing seam), MicroFreakDevice
// (port of device.py), and FreakDeviceFactory.
//
// The app holds `any FreakDeviceProtocol` and never learns whether it is
// hardware or the simulator. Every device write is read back and
// hash-verified by default; verify: false is the explicit per-call opt-out.

import Foundation

public struct SnapshotOptions: Sendable {
    public var readBlobs: Bool = true     // false = fast names-only pass (sha256 nil)
    public var keepBlobs: Bool = false    // true required for Library.importSnapshot
    public var slots: [Int]? = nil        // nil = all; sorted ascending before use

    public init() {}
}

public struct BackupOptions: Sendable {
    public var slots: [Int]? = nil
    public var resume: Bool = false       // skip slots already on disk with an index entry

    public init() {}
}

public protocol FreakDeviceProtocol: Sendable {
    var slotCount: Int { get }

    // reads
    func name(slot: Int) async throws -> String                       // ~1 ms on hardware
    func read(slot: Int) async throws -> Preset                       // ~400 ms on hardware
    func snapshot(options: SnapshotOptions,
                  progress: ProgressReporter?) async throws -> DeviceSnapshot

    // writes (verified by default via the extension overloads)
    func write(slot: Int, preset: Preset, verify: Bool) async throws -> WriteReport
    func rename(slot: Int, name: String, verify: Bool) async throws -> WriteReport

    // backup / restore
    func backup(to dest: URL, options: BackupOptions,
                progress: ProgressReporter?) async throws -> BackupSet
    func restore(from source: BackupSet, slots: [Int]?, verify: Bool,
                 progress: ProgressReporter?) async throws -> [WriteReport]

    func close() async
}

/// Default-argument overloads (protocols cannot declare defaults).
public extension FreakDeviceProtocol {
    func snapshot() async throws -> DeviceSnapshot {
        try await snapshot(options: SnapshotOptions(), progress: nil)
    }

    func snapshot(options: SnapshotOptions) async throws -> DeviceSnapshot {
        try await snapshot(options: options, progress: nil)
    }

    func write(slot: Int, preset: Preset) async throws -> WriteReport {
        try await write(slot: slot, preset: preset, verify: true)
    }

    func rename(slot: Int, name: String) async throws -> WriteReport {
        try await rename(slot: slot, name: name, verify: true)
    }

    func backup(to dest: URL) async throws -> BackupSet {
        try await backup(to: dest, options: BackupOptions(), progress: nil)
    }

    func restore(from source: BackupSet) async throws -> [WriteReport] {
        try await restore(from: source, slots: nil, verify: true, progress: nil)
    }
}

public final class MicroFreakDevice: FreakDeviceProtocol, Sendable {
    public let slotCount: Int
    public let session: FreakSession          // escape hatch, mirrors Python .session
    private let clock: any FreakClock

    public init(transport: any FreakTransport,
                slotCount: Int = Wire.slots,
                config: SessionConfig = SessionConfig(),
                clock: any FreakClock = SystemClock()) {
        self.slotCount = slotCount
        self.clock = clock
        self.session = FreakSession(transport: transport, config: config, clock: clock)
    }

    public func close() async {
        await session.close()
    }

    // ------------------------------------------------------------- reads

    public func name(slot: Int) async throws -> String {
        try await session.readName(slot: slot).name
    }

    public func read(slot: Int) async throws -> Preset {
        let info = try await session.readName(slot: slot)
        let blob = try await session.readBlob(slot: slot)
        return try Preset(name: info.name, blob: blob, meta: info.meta)
    }

    /// Read names (and, by default, blobs + hashes) for the requested slots.
    /// Cancellation throws .operationCancelled; no partial snapshot is ever
    /// returned. Name read failures of kind .deviceTimeout / .replyMismatch
    /// are swallowed (the record gets name nil, meta nil); everything else
    /// propagates.
    public func snapshot(options: SnapshotOptions,
                         progress: ProgressReporter?) async throws -> DeviceSnapshot {
        defer { progress?.finish() }
        let slotList = options.slots.map { $0.sorted() } ?? Array(0..<slotCount)
        let total = slotList.count
        var records: [SlotRecord] = []
        var nameMs: [Double] = []
        var dumpMs: [Double] = []
        var durations: [Double] = []
        let tStart = clock.now
        for (done, slot) in slotList.enumerated() {
            if Task.isCancelled {
                throw FreakError.operationCancelled(done: done, total: total)
            }
            let tSlot = clock.now
            var nm: String? = nil
            var meta: Data? = nil
            do {
                let t0 = clock.now
                let info = try await session.readName(slot: slot)
                nameMs.append((clock.now - t0) * 1000)
                nm = info.name
                meta = info.meta
            } catch let e as FreakError {
                switch e {
                case .deviceTimeout, .replyMismatch:
                    break    // name = nil, meta = nil on the record
                default:
                    throw e
                }
            }
            var sha: String? = nil
            var kept: Data? = nil
            if options.readBlobs {
                let t0 = clock.now
                let blob = try await session.readBlob(slot: slot)
                dumpMs.append((clock.now - t0) * 1000)
                sha = Wire.digest(blob)
                if options.keepBlobs {
                    kept = blob
                }
            }
            records.append(SlotRecord(slot: slot, name: nm, sha256: sha,
                                      meta: meta, blob: kept))
            durations.append(clock.now - tSlot)
            if let progress {
                let elapsed = clock.now - tStart
                let med = durations.sorted()[durations.count / 2]
                progress.report(ProgressEvent(
                    done: done + 1, total: total, slot: slot, name: nm ?? "",
                    elapsedSeconds: elapsed,
                    etaSeconds: med * Double(total - done - 1)))
            }
        }
        let elapsed = clock.now - tStart
        let timing = TimingReport(
            totalSeconds: roundTo(elapsed, places: 3),
            perSlotSeconds: roundTo(elapsed / Double(max(total, 1)), places: 4),
            nameMsMedian: median(nameMs),
            dumpMsMedian: median(dumpMs))
        return DeviceSnapshot(takenAt: isoNow(), records: records, timing: timing)
    }

    // ------------------------------------------ writes (verified by default)

    public func write(slot: Int, preset: Preset, verify: Bool) async throws -> WriteReport {
        let t0 = clock.now
        let info = try await session.writePreset(slot: slot, preset: preset)
        var verified: Bool? = nil
        if verify {
            if info.name != preset.name {
                throw FreakError.verifyMismatch(VerifyMismatch(
                    slot: slot, expectedSha256: preset.sha256, actualSha256: nil,
                    expectedName: preset.name, actualName: info.name,
                    firstDifference: nil, expectedLength: preset.blob.count,
                    actualLength: 0))
            }
            let blob = try await session.readBlob(slot: slot)
            let actual = Wire.digest(blob)
            if actual != preset.sha256 {
                let expected = [UInt8](preset.blob)
                let got = [UInt8](blob)
                let n = min(expected.count, got.count)
                let first = (0..<n).first { expected[$0] != got[$0] } ?? n
                throw FreakError.verifyMismatch(VerifyMismatch(
                    slot: slot, expectedSha256: preset.sha256, actualSha256: actual,
                    expectedName: preset.name, actualName: info.name,
                    firstDifference: first, expectedLength: expected.count,
                    actualLength: got.count))
            }
            verified = true
        }
        return WriteReport(slot: slot, sha256: preset.sha256, name: preset.name,
                           verified: verified, durationSeconds: clock.now - t0)
    }

    public func rename(slot: Int, name: String, verify: Bool) async throws -> WriteReport {
        try Wire.validateName(name)
        let t0 = clock.now
        let current = try await session.readName(slot: slot)          // current meta
        let info = try await session.writeName(slot: slot, name: name, meta: current.meta)
        var verified: Bool? = nil
        if verify {
            if info.name != name {
                throw FreakError.verifyMismatch(VerifyMismatch(
                    slot: slot, expectedSha256: "", actualSha256: nil,
                    expectedName: name, actualName: info.name,
                    firstDifference: nil, expectedLength: 0, actualLength: 0))
            }
            verified = true
        }
        return WriteReport(slot: slot, sha256: "", name: name,
                           verified: verified, durationSeconds: clock.now - t0)
    }

    // ------------------------------------------------------ backup / restore

    /// Read every requested slot to the phase-0 on-disk format, persisting
    /// as it goes — each slot is on disk before the next is read, so a
    /// cancelled pass leaves valid partial state. Reads only; never writes
    /// to the device.
    public func backup(to dest: URL, options: BackupOptions,
                       progress: ProgressReporter?) async throws -> BackupSet {
        defer { progress?.finish() }
        let fm = FileManager.default
        let presetsDir = dest.appendingPathComponent("presets")
        try fm.createDirectory(at: presetsDir, withIntermediateDirectories: true)
        let indexPath = dest.appendingPathComponent("index.json")
        var index: [String: Any]
        if fm.fileExists(atPath: indexPath.path) {
            let raw = try Data(contentsOf: indexPath)
            index = (try JSONSerialization.jsonObject(with: raw) as? [String: Any]) ?? [:]
            if index["presets"] == nil { index["presets"] = [String: Any]() }
            if index["timing"] == nil { index["timing"] = [String: Any]() }
        } else {
            index = ["created": isoNow(), "slots": slotCount,
                     "presets": [String: Any](), "timing": [String: Any]()]
        }
        var presets = (index["presets"] as? [String: Any]) ?? [:]

        let slotList = options.slots.map { $0.sorted() } ?? Array(0..<slotCount)
        let total = slotList.count
        var nameMs: [Double] = []
        var dumpMs: [Double] = []
        var durations: [Double] = []
        let tStart = clock.now
        for (done, slot) in slotList.enumerated() {
            if Task.isCancelled {
                throw FreakError.operationCancelled(done: done, total: total)
            }
            let binPath = presetsDir.appendingPathComponent(String(format: "%03d.bin", slot))
            if options.resume && fm.fileExists(atPath: binPath.path)
                && presets[String(slot)] != nil {
                continue
            }
            let tSlot = clock.now
            var t0 = clock.now
            let info = try await session.readName(slot: slot)  // NOT swallowed here —
            nameMs.append((clock.now - t0) * 1000)             // a backup wants to know
            t0 = clock.now
            let blob = try await session.readBlob(slot: slot)
            dumpMs.append((clock.now - t0) * 1000)
            try blob.write(to: binPath)
            presets[String(slot)] = ["slot": slot, "name": info.name,
                                     "bytes": blob.count,
                                     "sha256": Wire.digest(blob),
                                     "meta_hex": info.meta.hexString]
            index["presets"] = presets
            try AtomicFile.write(try jsonData(index), to: indexPath)
            durations.append(clock.now - tSlot)
            if let progress {
                let elapsed = clock.now - tStart
                let med = durations.sorted()[durations.count / 2]
                progress.report(ProgressEvent(
                    done: done + 1, total: total, slot: slot, name: info.name,
                    elapsedSeconds: elapsed,
                    etaSeconds: med * Double(total - done - 1)))
            }
        }
        let elapsed = clock.now - tStart
        let readCount = durations.count
        index["timing"] = [
            "total_seconds": roundTo(elapsed, places: 3),
            "per_slot_seconds": roundTo(elapsed / Double(max(readCount, 1)), places: 4),
            "name_ms_median": median(nameMs).map { $0 as Any } ?? NSNull(),
            "dump_ms_median": median(dumpMs).map { $0 as Any } ?? NSNull(),
        ] as [String: Any]
        try AtomicFile.write(try jsonData(index), to: indexPath)
        return try BackupSet.load(dest)
    }

    /// Write presets from a BackupSet back to the device. Stops at the first
    /// thrown error — a failing write path must not keep writing — and any
    /// FreakError (cancellation included) is rethrown as
    /// .restoreFailed(underlying:completed:).
    public func restore(from source: BackupSet, slots: [Int]?, verify: Bool,
                        progress: ProgressReporter?) async throws -> [WriteReport] {
        defer { progress?.finish() }
        let slotList = slots.map { $0.sorted() } ?? source.coveredSlots()
        let total = slotList.count
        var reports: [WriteReport] = []
        var durations: [Double] = []
        let tStart = clock.now
        for (done, slot) in slotList.enumerated() {
            let rep: WriteReport
            do {
                if Task.isCancelled {
                    throw FreakError.operationCancelled(done: done, total: total)
                }
                let preset = try source.preset(slot)
                let t0 = clock.now
                rep = try await write(slot: slot, preset: preset, verify: verify)
                durations.append(clock.now - t0)
            } catch let e as FreakError {
                throw FreakError.restoreFailed(underlying: e, completed: reports)
            }
            reports.append(rep)
            if let progress {
                let elapsed = clock.now - tStart
                let med = durations.sorted()[durations.count / 2]
                progress.report(ProgressEvent(
                    done: done + 1, total: total, slot: slot, name: rep.name,
                    elapsedSeconds: elapsed,
                    etaSeconds: med * Double(total - done - 1)))
            }
        }
        return reports
    }
}

// ----------------------------------------------------------- device factory

public enum FreakDeviceFactory {
    /// Practice mode: MicroFreakDevice over SimulatedMicroFreak.factoryFresh.
    /// Instant, offline, 269 duplicated Init slots among named pseudo-presets
    /// — the full UI (browse, backup, sync, expendability badges) works on
    /// the couch.
    public static func practice(initCopies: Int = 269, seed: Int = 0) -> any FreakDeviceProtocol {
        MicroFreakDevice(transport: SimulatedMicroFreak.factoryFresh(
            initCopies: initCopies, seed: seed))
    }

    #if canImport(CoreMIDI)
    /// Hardware: MicroFreakDevice over CoreMIDITransport.open(hints:exclude:).
    /// Throws .deviceNotFound / .transport — the connect screen shows the
    /// endpoint lists carried in .deviceNotFound and offers practice mode as
    /// the fallback.
    public static func hardware(hints: [String] = CoreMIDITransport.defaultHints)
        throws -> any FreakDeviceProtocol {
        MicroFreakDevice(transport: try CoreMIDITransport.open(hints: hints))
    }
    #endif
}

File: microfreak/device.py
"""MicroFreak: the librarian's device API.

Every device write is read back and hash-verified by default; verify=False
is the explicit per-call opt-out.
"""
from __future__ import annotations

import json
import time
from pathlib import Path
from typing import Callable, List, Optional, Sequence, Union

from .backup import BackupSet, atomic_write_text
from .errors import (DeviceTimeoutError, MicroFreakError,
                     OperationCancelledError, ReplyMismatchError,
                     VerifyMismatchError)
from .model import (CancelToken, DeviceSnapshot, Preset, ProgressEvent,
                    ProgressFn, SlotRecord, TimingReport, WriteReport)
from .protocol import SLOTS, digest, validate_name
from .session import Session
from .transport import Transport


def _median(xs: List[float]) -> Optional[float]:
    """Median as mfcap.midi.backup computes it: sorted()[len // 2]."""
    if not xs:
        return None
    return round(sorted(xs)[len(xs) // 2], 1)


def _iso_now() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%S")


class MicroFreak:
    def __init__(self, transport: Transport, *, slots: int = SLOTS,
                 clock: Callable[[], float] = time.monotonic,
                 sleep: Callable[[float], None] = time.sleep):
        self.slots = slots
        self.clock = clock
        self._session = Session(transport, clock=clock, sleep=sleep)

    @property
    def session(self) -> Session:
        return self._session

    def close(self) -> None:
        self._session.close()

    def __enter__(self) -> "MicroFreak":
        return self

    def __exit__(self, *exc) -> None:
        self.close()

    # ------------------------------------------------------------- reads

    def name(self, slot: int) -> str:
        return self._session.read_name(slot).name

    def read(self, slot: int) -> Preset:
        info = self._session.read_name(slot)
        blob = self._session.read_blob(slot)
        return Preset(name=info.name, blob=blob, meta=info.meta)

    def snapshot(self, *, read_blobs: bool = True, keep_blobs: bool = False,
                 slots: Optional[Sequence[int]] = None,
                 progress: Optional[ProgressFn] = None,
                 cancel: Optional[CancelToken] = None) -> DeviceSnapshot:
        """Read names (and, by default, blobs + hashes) for the requested
        slots. Cancellation raises OperationCancelledError; no partial
        snapshot is returned."""
        slot_list = sorted(slots) if slots is not None else list(range(self.slots))
        total = len(slot_list)
        records: List[SlotRecord] = []
        name_ms: List[float] = []
        dump_ms: List[float] = []
        durations: List[float] = []
        t_start = self.clock()
        for done, slot in enumerate(slot_list):
            if cancel is not None and cancel.cancelled:
                raise OperationCancelledError(done, total)
            t_slot = self.clock()
            nm: Optional[str] = None
            meta: Optional[bytes] = None
            try:
                t0 = self.clock()
                info = self._session.read_name(slot)
                name_ms.append((self.clock() - t0) * 1000)
                nm, meta = info.name, info.meta
            except (DeviceTimeoutError, ReplyMismatchError):
                pass    # name=None, meta=None on the record
            sha: Optional[str] = None
            kept: Optional[bytes] = None
            if read_blobs:
                t0 = self.clock()
                blob = self._session.read_blob(slot)
                dump_ms.append((self.clock() - t0) * 1000)
                sha = digest(blob)
                if keep_blobs:
                    kept = blob
            records.append(SlotRecord(slot=slot, name=nm, sha256=sha,
                                      meta=meta, blob=kept))
            durations.append(self.clock() - t_slot)
            if progress is not None:
                elapsed = self.clock() - t_start
                med = sorted(durations)[len(durations) // 2]
                eta = med * (total - done - 1) if durations else None
                progress(ProgressEvent(done=done + 1, total=total, slot=slot,
                                       name=nm or "", elapsed_seconds=elapsed,
                                       eta_seconds=eta))
        elapsed = self.clock() - t_start
        timing = TimingReport(
            total_seconds=round(elapsed, 3),
            per_slot_seconds=round(elapsed / max(total, 1), 4),
            name_ms_median=_median(name_ms),
            dump_ms_median=_median(dump_ms))
        return DeviceSnapshot(taken_at=_iso_now(), records=tuple(records),
                              timing=timing)

    # ------------------------------------------- writes (verified by default)

    def write(self, slot: int, preset: Preset, *, verify: bool = True,
              cancel: Optional[CancelToken] = None) -> WriteReport:
        t0 = self.clock()
        info = self._session.write_preset(slot, preset, cancel=cancel)
        verified: Optional[bool] = None
        if verify:
            if info.name != preset.name:
                raise VerifyMismatchError(
                    slot, preset.sha256, None, preset.name, info.name,
                    None, len(preset.blob), 0)
            blob = self._session.read_blob(slot)
            actual = digest(blob)
            if actual != preset.sha256:
                n = min(len(preset.blob), len(blob))
                first = next((i for i in range(n)
                              if preset.blob[i] != blob[i]), n)
                raise VerifyMismatchError(
                    slot, preset.sha256, actual, preset.name, info.name,
                    first, len(preset.blob), len(blob))
            verified = True
        return WriteReport(slot=slot, sha256=preset.sha256, name=preset.name,
                           verified=verified,
                           duration_seconds=self.clock() - t0)

    def rename(self, slot: int, name: str, *, verify: bool = True) -> WriteReport:
        validate_name(name)
        t0 = self.clock()
        current = self._session.read_name(slot)          # current meta
        info = self._session.write_name(slot, name, current.meta)
        verified: Optional[bool] = None
        if verify:
            if info.name != name:
                raise VerifyMismatchError(
                    slot, "", None, name, info.name, None, 0, 0)
            verified = True
        return WriteReport(slot=slot, sha256="", name=name, verified=verified,
                           duration_seconds=self.clock() - t0)

    # ------------------------------------------------------ backup / restore

    def backup(self, dest: Union[str, Path], *,
               slots: Optional[Sequence[int]] = None, resume: bool = False,
               progress: Optional[ProgressFn] = None,
               cancel: Optional[CancelToken] = None) -> BackupSet:
        """Read every requested slot to the phase-0 on-disk format, persisting
        as it goes (each slot written before the next is read, so a cancelled
        pass leaves valid partial state). Reads only; never writes to the
        device."""
        dest = Path(dest)
        presets_dir = dest / "presets"
        presets_dir.mkdir(parents=True, exist_ok=True)
        index_path = dest / "index.json"
        if index_path.exists():
            index = json.loads(index_path.read_text())
            index.setdefault("presets", {})
            index.setdefault("timing", {})
        else:
            index = {"created": _iso_now(), "slots": self.slots,
                     "presets": {}, "timing": {}}

        slot_list = sorted(slots) if slots is not None else list(range(self.slots))
        total = len(slot_list)
        name_ms: List[float] = []
        dump_ms: List[float] = []
        durations: List[float] = []
        t_start = self.clock()
        for done, slot in enumerate(slot_list):
            if cancel is not None and cancel.cancelled:
                raise OperationCancelledError(done, total)
            bin_path = presets_dir / f"{slot:03d}.bin"
            if resume and bin_path.exists() and str(slot) in index["presets"]:
                continue
            t_slot = self.clock()
            t0 = self.clock()
            info = self._session.read_name(slot)
            name_ms.append((self.clock() - t0) * 1000)
            t0 = self.clock()
            blob = self._session.read_blob(slot)
            dump_ms.append((self.clock() - t0) * 1000)
            bin_path.write_bytes(blob)
            index["presets"][str(slot)] = {
                "slot": slot, "name": info.name, "bytes": len(blob),
                "sha256": digest(blob), "meta_hex": info.meta.hex()}
            atomic_write_text(index_path, json.dumps(index, indent=2))
            durations.append(self.clock() - t_slot)
            if progress is not None:
                elapsed = self.clock() - t_start
                med = sorted(durations)[len(durations) // 2]
                progress(ProgressEvent(done=done + 1, total=total, slot=slot,
                                       name=info.name, elapsed_seconds=elapsed,
                                       eta_seconds=med * (total - done - 1)))
        elapsed = self.clock() - t_start
        read_count = len(durations)
        index["timing"] = {
            "total_seconds": round(elapsed, 3),
            "per_slot_seconds": round(elapsed / max(read_count, 1), 4),
            "name_ms_median": _median(name_ms),
            "dump_ms_median": _median(dump_ms)}
        atomic_write_text(index_path, json.dumps(index, indent=2))
        return BackupSet.load(dest)

    def restore(self, source: BackupSet,
                slots: Optional[Sequence[int]] = None, *, verify: bool = True,
                progress: Optional[ProgressFn] = None,
                cancel: Optional[CancelToken] = None) -> List[WriteReport]:
        """Write presets from a BackupSet back to the device. Stops at the
        first raised error (a failing write path must not keep writing);
        reports for completed slots are attached to the exception as
        .completed when it is a MicroFreakError."""
        slot_list = sorted(slots) if slots is not None else source.covered_slots()
        total = len(slot_list)
        reports: List[WriteReport] = []
        durations: List[float] = []
        t_start = self.clock()
        for done, slot in enumerate(slot_list):
            try:
                if cancel is not None and cancel.cancelled:
                    raise OperationCancelledError(done, total)
                preset = source.preset(slot)
                t0 = self.clock()
                rep = self.write(slot, preset, verify=verify, cancel=cancel)
                durations.append(self.clock() - t0)
            except MicroFreakError as e:
                e.completed = reports        # type: ignore[attr-defined]
                raise
            reports.append(rep)
            if progress is not None:
                elapsed = self.clock() - t_start
                med = sorted(durations)[len(durations) // 2]
                progress(ProgressEvent(done=done + 1, total=total, slot=slot,
                                       name=rep.name, elapsed_seconds=elapsed,
                                       eta_seconds=med * (total - done - 1)))
        return reports
